### PR TITLE
Chunking

### PR DESCRIPTION
**Pull request by jsarabia**
This change is [Reviewable](https://reviewable.io/reviews/bible-translation-tools/orature/587)

### CODE REVIEW

**Comment by AnonymousWalker** [comment]
*[`common/src/main/kotlin/org/wycliffeassociates/otter/common/data/workbook/Chapter.kt` line 35 at r3](https://reviewable.io/reviews/bible-translation-tools/orature/587#-N4bkYm0BJib5W-DRZJV:-N4bkYm19qMcB4Lm1xa4:bpbj9h2) ([raw file](https://github.com/bible-translation-tools/orature/blob/ce6a6a78dbbcb6c5aa4ecb095fbb6e1ddc260200/common/src/main/kotlin/org/wycliffeassociates/otter/common/data/workbook/Chapter.kt#L35)):*
> ```Kotlin
> import org.wycliffeassociates.otter.common.domain.content.Recordable
> 
> data class Chapter(
> ```

this should be a class, as it should have been. Also, could the constructor parameters (chunks, chunkCount, addChunk, reset) go to another class /wrapper? I see a lot specific pieces that are not strongly related to a Chapter other than chunks.

**Comment by AnonymousWalker** [comment]
*[`common/src/main/kotlin/org/wycliffeassociates/otter/common/data/workbook/Chunk.kt` line 68 at r3](https://reviewable.io/reviews/bible-translation-tools/orature/587#-N4blWzw-S9UurvjqQAa:-N4blWzw-S9UurvjqQAb:b-9cnsog) ([raw file](https://github.com/bible-translation-tools/orature/blob/ce6a6a78dbbcb6c5aa4ecb095fbb6e1ddc260200/common/src/main/kotlin/org/wycliffeassociates/otter/common/data/workbook/Chunk.kt#L68)):*
> ```Kotlin
>         if (start != other.start) return false
>         if (end != other.end) return false
>         if (draftNumber != other.draftNumber) return false
> ```

you may consider using `when { }`

**Comment by AnonymousWalker** [comment]
*[`common/src/main/kotlin/org/wycliffeassociates/otter/common/domain/collections/CreateProject.kt` line 41 at r3](https://reviewable.io/reviews/bible-translation-tools/orature/587#-N4bm-XYB3foZdyLDPxl:-N4bm-XYB3foZdyLDPxm:b-p3933j) ([raw file](https://github.com/bible-translation-tools/orature/blob/ce6a6a78dbbcb6c5aa4ecb095fbb6e1ddc260200/common/src/main/kotlin/org/wycliffeassociates/otter/common/domain/collections/CreateProject.kt#L41)):*
> ```Kotlin
>         targetLanguage: Language,
>         resourceId: String? = null,
>         verseByVerse: Boolean = false
> ```

so we imply the default mode is chunk instead?

**Comment by AnonymousWalker** [comment]
*[`common/src/main/kotlin/org/wycliffeassociates/otter/common/domain/content/FileNamer.kt` line 77 at r3](https://reviewable.io/reviews/bible-translation-tools/orature/587#-N4bmHkQ6jVXCnIMB5iW:-N4bmHkQ6jVXCnIMB5iX:b-atfndc) ([raw file](https://github.com/bible-translation-tools/orature/blob/ce6a6a78dbbcb6c5aa4ecb095fbb6e1ddc260200/common/src/main/kotlin/org/wycliffeassociates/otter/common/domain/content/FileNamer.kt#L77)):*
> ```Kotlin
>         val verseNum = when (start) {
>             null -> null
>             else -> verseFormat.format(sort)
> ```

How about this?
`val verseNum = start?.let { verseFormat.format(sort) }`

**Comment by AnonymousWalker** [comment]
*[`common/src/main/kotlin/org/wycliffeassociates/otter/common/domain/content/VerseByVerseChunking.kt` line 35 at r3](https://reviewable.io/reviews/bible-translation-tools/orature/587#-N4bwjl69EvB3QEFzovu:-N4bwjl69EvB3QEFzovv:b-63j9r3) ([raw file](https://github.com/bible-translation-tools/orature/blob/ce6a6a78dbbcb6c5aa4ecb095fbb6e1ddc260200/common/src/main/kotlin/org/wycliffeassociates/otter/common/domain/content/VerseByVerseChunking.kt#L35)):*
> ```Kotlin
> 
>     init {
>         accessor = ProjectFilesAccessor(
> ```

Could we pass a ProjectFilesAccessor from constructor instead of re-creating a new one here?

**Comment by AnonymousWalker** [comment]
*[`common/src/main/kotlin/org/wycliffeassociates/otter/common/domain/resourcecontainer/projectimportexport/ProjectImporter.kt` line 195 at r3](https://reviewable.io/reviews/bible-translation-tools/orature/587#-N4c3YT_AaP3QWRbNH1m:-N4c3YT_AaP3QWRbNH1n:b-fyphz4) ([raw file](https://github.com/bible-translation-tools/orature/blob/ce6a6a78dbbcb6c5aa4ecb095fbb6e1ddc260200/common/src/main/kotlin/org/wycliffeassociates/otter/common/domain/resourcecontainer/projectimportexport/ProjectImporter.kt#L195)):*
> ```Kotlin
>     }
> 
>     private fun importChunks(project: Collection, accessor: ProjectFilesAccessor, fileReader: IFileReader) {
> ```

needs refactor to smaller functions

**Comment by AnonymousWalker** [comment]
*[`jvm/workbookapp/src/main/kotlin/org/wycliffeassociates/otter/jvm/workbookapp/ui/screens/chunking/Consume.kt` line 37 at r3](https://reviewable.io/reviews/bible-translation-tools/orature/587#-N4cQ-Yh1kBuurZWtDNT:-N4cQ-Yh1kBuurZWtDNU:b-s9b271) ([raw file](https://github.com/bible-translation-tools/orature/blob/ce6a6a78dbbcb6c5aa4ecb095fbb6e1ddc260200/jvm/workbookapp/src/main/kotlin/org/wycliffeassociates/otter/jvm/workbookapp/ui/screens/chunking/Consume.kt#L37)):*
> ```Kotlin
> import tornadofx.*
> 
> class Consume : Fragment() {
> ```

using fragment could lead to memory leak if the view contains images

**Review by AnonymousWalker — CHANGES_REQUESTED** [review]
> Reviewed 33 of 59 files at r1, 28 of 29 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/bible-translation-tools/orature/587)* status: 61 of 69 files reviewed, 7 unresolved discussions (waiting on @AnonymousWalker and @jsarabia)

File: common/src/main/kotlin/org/wycliffeassociates/otter/common/data/workbook/Chapter.kt [hunk 1]
```diff
@@ -32,26 +38,48 @@ data class Chapter(
...
+        return Single.fromCallable {
+            val draft = mutableListOf<Chunk>()
+            val chunkTotal = chunkCount.blockingGet()
+            val disposable = chunks
+                .filter { it.draftNumber > 0 } // filter active drafts
+                .takeUntil { draft.size == chunkTotal }
+                .forEach { draft.add(it) }
+            while (draft.size != chunkTotal) {
+                sleep(50)
+            }
```
Comment (jsarabia) [inline common/src/main/kotlin/org/wycliffeassociates/otter/common/data/workbook/Chapter.kt:68]: explain why this is necessary

**Review by AnonymousWalker — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/bible-translation-tools/orature/587)* status: 17 of 112 files reviewed, 9 unresolved discussions (waiting on @AnonymousWalker and @jsarabia)

___
*[`jvm/controls/src/main/kotlin/org/wycliffeassociates/otter/jvm/controls/waveform/IMarkerViewModel.kt` line 14 at r14](https://reviewable.io/reviews/bible-translation-tools/orature/587#-N96vIXM9gxZU_ii1tNB:-N96vIXM9gxZU_ii1tNC:bd7i6b7) ([raw file](https://github.com/bible-translation-tools/orature/blob/4e648ca7a65c05f9e61e981ad13b817fac5c90b1/jvm/controls/src/main/kotlin/org/wycliffeassociates/otter/jvm/controls/waveform/IMarkerViewModel.kt#L14)):*
> ```Kotlin
> import org.wycliffeassociates.otter.jvm.controls.model.VerseMarkerModel
> 
> interface IMarkerViewModel {
> ```

Just a reminder, this default implemented interface needs documentation to avoid confusion with its purpose.

**Review by mXaln — COMMENTED** [review]
> Reviewed 1 of 59 files at r1, 1 of 29 files at r3, 5 of 13 files at r5, 13 of 29 files at r6, 10 of 32 files at r8, 14 of 26 files at r9, 9 of 18 files at r10, 1 of 16 files at r11, 14 of 24 files at r12, 8 of 11 files at r13, 16 of 20 files at r15, 3 of 3 files at r16, 2 of 2 files at r17, 1 of 2 files at r18, 3 of 3 files at r19, 5 of 5 files at r20, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/bible-translation-tools/orature/587)* status: all files reviewed, 11 unresolved discussions (waiting on @AnonymousWalker, @jsarabia, and @mXaln)

___
*[`common/src/main/kotlin/org/wycliffeassociates/otter/common/data/workbook/Chapter.kt` line 41 at r9](https://reviewable.io/reviews/bible-translation-tools/orature/587#-N7LNfUP7rfd30M8RnWU:-N7LNfUP7rfd30M8RnWV:b-t2vvlz) ([raw file](https://github.com/bible-translation-tools/orature/blob/783b6c03810feba9b0aa0e4ff95ecb3158bef298/common/src/main/kotlin/org/wycliffeassociates/otter/common/data/workbook/Chapter.kt#L41)):*
> ```Kotlin
>     override val resources: List<ResourceGroup>,
>     override val subtreeResources: List<ResourceMetadata>,
>     private val lazychunks: Lazy<ReplayRelay<Chunk>>,
> ```

Maybe make it camelCase?
___
*[`common/src/main/kotlin/org/wycliffeassociates/otter/common/persistence/repositories/WorkbookRepository.kt` line 231 at r19](https://reviewable.io/reviews/bible-translation-tools/orature/587#-N9aAro9C91r36qVscSO:-N9aAro9C91r36qVscSP:b-ahstsa) ([raw file](https://github.com/bible-translation-tools/orature/blob/71d30ccfeec623ab4edc8eda2be4ec39bec491cf/common/src/main/kotlin/org/wycliffeassociates/otter/common/persistence/repositories/WorkbookRepository.kt#L231)):*
> ```Kotlin
>     ): ReplayRelay<Chunk> {
>         val rr = ReplayRelay.create<Chunk>()
>         db.getContentByCollectionActiveConnection(chapterCollection)
> ```

This name is too long and I don't quite understand what Active Connection means.

**Pull request closed without merging**